### PR TITLE
Aj/ora2 dsl

### PR DESCRIPTION
**Pull request by staubina**
### Description

Adds a job to the ora2 directory to create the SetORS2VersionOnSandbox jenkins job.

WIP: This is still a work in progress.  @jibsheet Please take a look when you have a moment.  I want to make sure the approach to the RunAnsibleRemote is appropriate.  I removed what I was not using for parameters.  I have tested this job locally and it is working.

- [x] @jibsheet 
- [x] @jlajoie please review

**FYI**
@cahrens 

- [ ] Rebase and squash commits

### CODE REVIEW

File: ora2/jobs/SetOra2VersionOnSandbox.groovy [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+class SetOra2VersionOnSandbox {
+
+    public static def job = { dslFactory, extraVars ->
+        def jobRoot = extraVars.get("JOB_FOLDER_NAME")
+
+        def jobName = 'SetOra2VersionOnSandbox'
+        if (jobRoot != null && jobRoot != '') {
+            jobName = jobRoot + '/SetOra2VersionOnSandbox'
+        }
+        return dslFactory.job(jobName) {
```
Comment (jlajoie) [inline ora2/jobs/SetOra2VersionOnSandbox.groovy]: So I think you'll probably always want to have a folder name specified. Also, I used a get with a default, and I think the way I have it might look a bit cleaner? 

`dslFactory.job(extraVars.get('JOB_FOLDER_NAME', 'ora2') + '/some-job-name')`

Let me know what your opinion, but I think we should be consistent with each other.

File: ora2/jobs/SetOra2VersionOnSandbox.groovy [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            properties {
+                rebuild {
```
Comment (jlajoie) [inline ora2/jobs/SetOra2VersionOnSandbox.groovy]: So Kevin commented this on mine, it typically looks like we don't need these explicitly declared if they're just the defaults used, so unless we're overriding any settings we can omit this tag.

File: ora2/jobs/SetOra2VersionOnSandbox.groovy (hunk 1 above)
Comment (staubina) [inline ora2/jobs/SetOra2VersionOnSandbox.groovy]: This is mainly needed to that I can continue to test locally without a folder.  I do not think a folder is required, just helpful.

File: devops/jobs/RunRemoteAnsiblePlaybook.groovy [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            wrappers {
+              sshAgent(extraVars.get("SSH_AGENT_KEY", ''))
+            }
+
+            parameters {
+                stringParam('CONFIGURATION_REPO', extraVars.get('CONFIGURATION_REPO', 'https://github.com/edx/configuration.git'),
+                            'Git repo containing the analytics pipeline configuration automation.')
+                stringParam('CONFIGURATION_BRANCH', extraVars.get('CONFIGURATION_BRANCH', 'master'),
+                            'e.g. tagname or origin/branchname')
+                stringParam("SANDBOX_HOST",
```
Comment (jibsheet) [inline devops/jobs/RunRemoteAnsiblePlaybook.groovy]: I realize this is probably from your testing, but for a generic job class of RunRemoveAnsiblePlaybook, this should be INVENTORY (will want to update the other params too)

File: devops/jobs/RunRemoteAnsiblePlaybook.groovy [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                            "staubina-ora2.sandbox.edx.org",
+                            "The host name and address of the Sandbox to be updated")
+                stringParam("ORA2_VERSION",
```
Comment (jibsheet) [inline devops/jobs/RunRemoteAnsiblePlaybook.groovy]: Same thing, this can't have ORA2 inside a reusable piece of code

File: devops/jobs/RunRemoteAnsiblePlaybook.groovy [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+                           "test2",
+                           "The host name and address of the Sandbox to be updated")
+                stringParam("NOTIFY_ON_FAILURE",
+                            "astaubin@edx.org",
```
Comment (jibsheet) [inline devops/jobs/RunRemoteAnsiblePlaybook.groovy]: Presumably from testing, but should default to ''

File: devops/jobs/RunRemoteAnsiblePlaybook.groovy [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            triggers {
+                if (extraVars.get('SCHEDULE')) {
+                    cron(extraVars['SCHEDULE'])
+                }
+            }
+            throttleConcurrentBuilds {
+                maxPerNode(1)
+            }
+
+            def ora2_version_str = extraVars.get("ORA2_VERSION", "test3")
```
Comment (jibsheet) [inline devops/jobs/RunRemoteAnsiblePlaybook.groovy]: Similar comment from above

File: devops/jobs/RunRemoteAnsiblePlaybook.groovy [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+            // Parameters from the seed job that you don't override, they're just part of
+            // the run.  We need to load different ssh keys per env anyway.
+            environmentVariables {
+                env('PLAYBOOK', extraVars.get('PLAYBOOK'))
+                env('ANSIBLE_INVENTORY', extraVars.get('SANDBOX_HOST', ''))
+                env("REGION", extraVars.get('REGION','us-east-1'))
+                env("ANSIBLE_SSH_USER", extraVars.get('SSH_USER','ubuntu'))
+                env('ANSIBLE_EXTRA_VARS', "-e ora2_version=${ora2_version_str}")
```
Comment (jibsheet) [inline devops/jobs/RunRemoteAnsiblePlaybook.groovy]: This is something you'll want to construct in the seed job, not in a generic task

File: ora2/jobs/SetOra2VersionOnSandbox.groovy (hunk 1 above)
Comment (jibsheet) [inline ora2/jobs/SetOra2VersionOnSandbox.groovy]: While testing helpful, but for deployment this should really always have a folder

File: ora2/jobs/SetOra2VersionOnSandbox.groovy [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    PLAYBOOK:  The playbook used in this job
+*/
+
+package ora2.jobs
+
+import static devops.jobs.RunRemoteAnsiblePlaybook.job as RunRemoteAnsiblePlaybookJob
+class SetOra2VersionOnSandbox {
+
+    public static def job = { dslFactory, extraVars ->
+        def jobRoot = extraVars.get("JOB_FOLDER_NAME")
```
Comment (jibsheet) [inline ora2/jobs/SetOra2VersionOnSandbox.groovy]: Reasoning for the terminology change?  All the others are FOLDER_NAME

File: ora2/jobs/SetOra2VersionOnSandbox.groovy [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+                            "astaubin@edx.org",
+                            "Email to notify on failure")
+            }
+
+            concurrentBuild(false)
+            def ansibleJobName = 'ansibleOraVersion'
+            if (jobRoot != null && jobRoot != '') {
+                ansibleJobName = jobRoot + '/ansibleOraVersion'
+            }
+            RunRemoteAnsiblePlaybookJob(dslFactory, ansibleJobName, 'development', 'edx', extraVars + ['ORA2_VERSION':'$ORA2_VERSION', 'SANDBOX_HOST':'$SANDBOX_HOST'] )
```
Comment (jibsheet) [inline ora2/jobs/SetOra2VersionOnSandbox.groovy]: This looks like you're set up to pass down the parameters you want, so I'm confused why the ORA2 things were in RunRemoteAnsiblePlaybookJob

File: ora2/jobs/SetOra2VersionOnSandbox.groovy (hunk 8 above)
Comment (staubina) [inline ora2/jobs/SetOra2VersionOnSandbox.groovy]: The seed job used this name I will work with @jlajoie

File: ora2/jobs/SetOra2VersionOnSandbox.groovy (hunk 1 above)
Comment (staubina) [inline ora2/jobs/SetOra2VersionOnSandbox.groovy]: The issue is that I cannot test at all if I remove this or name my ansible job anything different.  This does not remove the ability to create with a folder, it just adds the ability to create the job when a folder is not present.  

Before merging I can remove this, but I will be unable to test it.

**Comment by staubina** [comment]
@jibsheet @jlajoie Please ignore this PR, while I test the requested changes.  I will squash and let you know when it is ready to be reviewed again.

**Comment by staubina** [comment]
@jibsheet @jlajoie I have made all the requested changes except the addition of a required folder.  This will have to be added later just before merging.

**Comment by staubina** [comment]
@jibsheet @jlajoie I rebased and set the PR to master.  Please review when you have a chance.

**Comment by jlajoie** [comment]
I think we're probably okay to delete the .gitkeep files at this point.

File: ora2/jobs/SetOra2VersionOnSandbox.groovy [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+package ora2.jobs
+
+class SetOra2VersionOnSandbox {
+
+    public static def job = { dslFactory, extraVars ->
+        def jobRoot = extraVars.get("FOLDER_NAME")
+
+        def jobName = 'SetOra2VersionOnSandbox'
+        if (jobRoot != null && jobRoot != '') {
+            jobName = jobRoot + '/SetOra2VersionOnSandbox'
```
Comment (jlajoie) [inline ora2/jobs/SetOra2VersionOnSandbox.groovy]: Is this still a placeholder for the issues you're running into on localdev?

File: devops/resources/run-remote-ansible-playbook.sh [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+  ANSIBLE_TAGS=""
+fi
+
+if [ -n "${ANSIBLE_INVENTORY}" ]; then
+    ANSIBLE_INVENTORY="${ANSIBLE_INVENTORY}"
+else
+    echo "You must specify ANSIBLE_INVENTORY"
+    exit
+fi
+ssh-add -l
```
Comment (jibsheet) [inline devops/resources/run-remote-ansible-playbook.sh]: debugging

File: devops/resources/run-remote-ansible-playbook.sh [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+# Test version works with existing Username
```
Comment (jibsheet) [inline devops/resources/run-remote-ansible-playbook.sh]: Also a debugging comment?

File: devops/jobs/RunRemoteAnsiblePlaybook.groovy [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+              sshAgent(extraVars.get("SSH_AGENT_KEY", ''))
+            }
+
+            parameters {
+                stringParam('CONFIGURATION_REPO', extraVars.get('CONFIGURATION_REPO', 'https://github.com/edx/configuration.git'),
+                            'Git repo containing the analytics pipeline configuration automation.')
+                stringParam('CONFIGURATION_BRANCH', extraVars.get('CONFIGURATION_BRANCH', 'master'),
+                            'e.g. tagname or origin/branchname')
+                stringParam("ANSIBLE_INVENTORY",
+                            "", "The inventory address to run the remote script")
```
Comment (jibsheet) [inline devops/jobs/RunRemoteAnsiblePlaybook.groovy]: Maybe reword as "The inventory address(es) where we run the remote script"  (You can provide multiple hosts to an inventory, hence my weird (es) phrasing).

File: devops/jobs/RunRemoteAnsiblePlaybook.groovy [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+                stringParam("ANSIBLE_EXTRA_VARS",
+                            "", "The Extra variables for Ansible")
```
Comment (jibsheet) [inline devops/jobs/RunRemoteAnsiblePlaybook.groovy]: You now know what this looks like, do you want to document that this consumes "multiple -e 'key=value' options" ?

File: ora2/jobs/SetOra2VersionOnSandbox.groovy (hunk 10 above)
Comment (jibsheet) [inline ora2/jobs/SetOra2VersionOnSandbox.groovy]: Same question

Comment (jibsheet) [inline ora2/jobs/SetOra2VersionOnSandbox.groovy]: Oh, it might be in order to do the dynamic calculations below of the downstream job.

File: ora2/jobs/SetOra2VersionOnSandbox.groovy [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+        return dslFactory.job(jobName) {
+            parameters {
+
+                stringParam('ORA2_VERSION',
+                            'master',
+                            'The version/branch name of ORA2 to fetch/checkout on target machine'
+                            )
+                stringParam("SANDBOX_HOST",
+                            "staubina-ora2.sandbox.edx.org",
```
Comment (jibsheet) [inline ora2/jobs/SetOra2VersionOnSandbox.groovy]: If we're going to have a default, it should be not your sandbox, and should really get passed in from the seed job as an extraVar

File: ora2/jobs/SetOra2VersionOnSandbox.groovy [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+                            "The host name and address of the Sandbox to be updated"
+                           )
+                stringParam("NOTIFY_ON_FAILURE",
+                            "",
```
Comment (jibsheet) [inline ora2/jobs/SetOra2VersionOnSandbox.groovy]: This should be reading from an extraVar

**Comment by jibsheet** [comment]
I'd like to come up with a design where we don't need the downstream job, but that's on me not you.

File: ora2/jobs/SetOra2VersionOnSandbox.groovy (hunk 10 above)
Comment (staubina) [inline ora2/jobs/SetOra2VersionOnSandbox.groovy]: This is used to build the job name with or without a folder name being provided.  Yes it was a work around to deal with local dev issues.  I will remove if necessary and ask @jlajoie to test locally.

**Comment by staubina** [comment]
@jibsheet I have made the requested changes.  Please review again when you have a moment.

@jlajoie has tested for me locally and everything is working as expected.
